### PR TITLE
Add support for re-exporting types from other modules

### PR DESCRIPTION
**Pull request by tronical**
The `export * from "./other_file.slint"` syntax takes all the exports from that file and re-exports them. This simplifies our style wrapper code.

As a bonus this also cleans up some of the export handling in the compiler.

### CODE REVIEW

File: internal/compiler/tests/syntax/imports/import_errors.slint [hunk 1]
```diff
@@ -10,7 +10,7 @@ import { NotExported } from "../../typeloader/incpath/local_helper_type.slint";
 import { Nothing } from "";
 //                      ^error{Unexpected empty import url}
 
-import "invalid_export.slint";
+import "../../typeloader/incpath/local_helper_type.slint";
```
Comment (tronical) [inline internal/compiler/tests/syntax/imports/import_errors.slint:13]: This seemingly unrelated change was done to avoid an early error-out due to the file not existing. We're not interested in tracking the error of a non-existent file, we're interested in tracking the error when not import names are specified.

File: internal/compiler/parser/document.rs [hunk 2]
```diff
@@ -163,6 +164,24 @@ fn parse_export(p: &mut impl Parser) -> bool {
         }
     } else if p.peek().as_str() == "struct" {
         parse_struct_declaration(&mut *p)
+    } else if p.peek().as_str() == "*" {
+        let mut p = p.start_node(SyntaxKind::ExportModule);
+        p.consume(); // *
```
Comment (ogoffart) [inline internal/compiler/parser/document.rs:169]: could be an `expect()`

**Review by ogoffart — COMMENTED** [review]
Missing:
 - [ ] changelog entry
 - [ ] documentation

Good change.

I think we also should have that for `import * from "..."`.
And I wonder if there should not be a way to be more resilient to duplicate so that adding an export is not a breaking change.
For example, forbidding the use of two *,  or always prioritizing the last one. or maybe the last one declared. Or the one for which the path is "closer"  or we need some kind of versioning.

File: internal/compiler/parser/document.rs (hunk 2 above)
Comment (tronical) [reply internal/compiler/parser/document.rs:169]: I've changed it to use `SyntaxKind::Star` instead.

**Comment by tronical** [comment]
> Missing:
> 
> * [ ]  changelog entry
> * [ ]  documentation

Done :)

> Good change.
> 
> I think we also should have that for `import * from "..."`. And I wonder if there should not be a way to be more resilient to duplicate so that adding an export is not a breaking change. For example, forbidding the use of two *, or always prioritizing the last one. or maybe the last one declared. Or the one for which the path is "closer" or we need some kind of versioning.

I'm also a bit unsure what the best behavior is, hence the strictest for now and we can relax later.

**Comment by ogoffart** [comment]
I think this change is good.  But right now, it is possible to di

```js
export * from "A.slint";
export * from "B.slint";
```

And if A or B add something that have something in common, it will start breaking.
So maybe we should forbid importing * from two different file?

Another idea is to force some kind of versioning when importing

```
import * from "A.slint" 1.0;
```

But the QML story told me this isn't great.

**Comment by tronical** [comment]
> I think this change is good. But right now, it is possible to di
> 
> ```js
> export * from "A.slint";
> export * from "B.slint";
> ```
> 
> And if A or B add something that have something in common, it will start breaking. So maybe we should forbid importing * from two different file?

I'd prefer the current behavior of showing an error if there is a conflict, as opposed to categorically forbidding it. But if you feel a little stronger about it, then we can also start with allowing re-exports only once per .slint file and show an error when used multiple times.

> Another idea is to force some kind of versioning when importing
> 
> ```
> import * from "A.slint" 1.0;
> ```
> 
> But the QML story told me this isn't great.

Yeah I don't think that we should go down that route :-)

**Comment by ogoffart** [comment]
> I'd prefer the current behavior of showing an error if there is a conflict, as opposed to categorically forbidding it.

That'd make adding any widget or component a breaking change technicly.
Can we ignore it?  
If we forbid two star import and give priority to the explicit ones, this wouldn't be a breaking change.

**Comment by tronical** [comment]
> That'd make adding any widget or component a breaking change technicly. Can we ignore it? If we forbid two star import and give priority to the explicit ones, this wouldn't be a breaking change.

Ok, I'll make a change that disallows multiple re-exports, prints a warning if a re-export clashes with a local export and priorities the local export of that type.

**Review by ogoffart — APPROVED** [review]
